File: aulas/20151005/01-Executors/SyncQueue.java
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.locks.*;

public class SyncQueue {
   
   private final LinkedList<Object> list =
      new LinkedList<>();
   
   public synchronized void put(Object obj) {
      list.add(obj);
      notify();
   }
   
   public synchronized Object take() throws InterruptedException {
      while (list.size() == 0) {
         wait();
      }
      return list.remove();
   }

   public static void main(String[] args) throws Exception {
      
      SyncQueue queue = new SyncQueue();
      
      Thread p1 = new Thread(() -> {
         for (int i = 0; i < 1024; ++i) {
            queue.put(new Object());
         }
      });
      
      Thread c1 = new Thread(() -> {
         for (int i = 0; i < 1024; ++i) {
            try {
               queue.take();
            } catch (Exception e) {
               System.err.println("Interrupt not supported!");
               System.exit(-1);
            }
         }
      });
      
      c1.start();
      p1.start();
      
      p1.join();
      c1.join();
      
      System.out.println("OK");
   }
   
}

File: aulas/20151001/SyncQueue.cs
using System;
using System.Threading;
using System.Collections.Generic;
using System.Runtime.CompilerServices;

public class SyncQueue {
   
   private readonly LinkedList<Object> list =
      new LinkedList<Object>();

   [MethodImpl(MethodImplOptions.Synchronized)]
   public void Put(Object obj) {
      list.AddLast(obj);
      Monitor.Pulse(this);
   }
   
   public Object Take() {
      lock (this) {
         while (list.Count == 0) {
            Monitor.Wait(this);
         }
         Object retval = list.First;
         list.RemoveFirst();
         return retval;
      }
   }

   public static void Main() {
      
      SyncQueue queue = new SyncQueue();
      
      Thread p1 = new Thread(() => {
         for (int i = 0; i < 1024; ++i) {
            queue.Put(new Object());
         }
      });
      
      Thread c1 = new Thread(() => {
         for (int i = 0; i < 1024; ++i) {
            queue.Take();
         }
      });
      
      c1.Start();
      p1.Start();
      
      p1.Join();
      c1.Join();
      
      Console.WriteLine("OK");
   }   
}
